src/darwin/Framework/CHIP/MTRAsyncWorkQueue.mm

```diff
@@ -264,27 +340,35 @@ - (void)_callNextReadyWorkItem
         while (_items.count >= 2) {
             MTRAsyncWorkItem * nextWorkItem = _items[1];
             if (!nextWorkItem.batchingHandler || nextWorkItem.batchingID != workItem.batchingID) {
-                break; // next item is not eligible to merge with this one
+                goto done; // next item is not eligible to merge with this one
             }
 
-            BOOL fullyMerged = NO;
-            batchingHandler(workItem.batchableData, nextWorkItem.batchableData, &fullyMerged);
-            if (!fullyMerged) {
-                break; // not removing the next item, so we can't merge anything else
+            switch (batchingHandler(workItem.batchableData, nextWorkItem.batchableData)) {
+            case MTRNotBatched:
+                goto done; // can't merge anything else
+            case MTRBatchedPartially:
+                MTR_LOG_DEFAULT("MTRAsyncWorkQueue<%@> partially merged work item %tu into %tu",
+                    context.description, nextWorkItem.uniqueID, workItem.uniqueID);
+                goto done; // can't merge anything else
+            case MTRBatchedFully:
+                MTR_LOG_DEFAULT("MTRAsyncWorkQueue<%@> fully merged work item %tu into %tu",
+                    context.description, nextWorkItem.uniqueID, workItem.uniqueID);
+                [_items removeObjectAtIndex:1];
+                continue; // try to batch the next item (if any)
```

Comment:
So if we are doing an explicit continue.... can't we just break out of the while loop after the switch, and just have normal `break` in all the places where we have `goto done;` right now?  

I guess either way is OK; just bothered by combining goto with languages that have things that care about scopes.  ;)

Reply:
I can do the extra breaks, but the goto seemed clearer in this case. `goto` has the well-defined expected behavior in the standard if you're jumping out of blocks (and not e.g. past initializers)

Reply:
Right, it's the accidental jumping past initializers that always worries me...  But in this case it's probably fine, like I said.